src/firewall/core/fw.py

```diff
@@ -555,21 +555,18 @@ def _start_apply_direct_rules(self):
     def _start_check(self):
         # check minimum required zones
         for z in [ "block", "drop", "trusted" ]:
-            if z not in self.zone.get_zones():
+            if self.zone.get_zone(z, required=False) is None:
```

Comment:
Yes, the goal is to avoid the O(n) lookup. It's a matter of principle.

The underscore indicates that `self.zone._zones` is private API of `fw_zone.py`. Currently, `src/firewall/core/fw.py` does not access that attribute (for correct reasons). On the other hand, `FirewallZone.get_zone()` is internal API that is accessed by other modules. We should strive for such separation.

Reply:
> The underscore indicates that self.zone._zones is private API of fw_zone.py.

I'm not super concerned about that. If you are then add a new getter function, `get_zones_dict()`. It would minimize a lot of these changes.

Reply:
a `self.zone.get_zones_dist()` exposes the internals almost as much as directly using `self.zone._zones`.

A `if zone_name in self.zone.get_zones_dist()` does not perform the same checks that a `if self.zone.get_zone(zone_name, required=False)` does. Which, currently that is the fallback of an empty zone-name to `get_default_zone()`, but theoretically, it could be more.

There are 3 functions that work together, and do  something similar: `check_zone()`, `get_zone()`, `get_zones()`.

For example, this is gonna work
```
    for zone_name in self.zone.get_zones():
        assert self.zone.get_zone(zone_name)
        assert self.zone.check_zone(zone_name) == zone_name
```
because they encapsulate the same underlying semantics about zones. Exposing the internal dictionary does not achieve that.

If you dislike `if self.zone.get_zone(z, required=False)` from an aesthetical POV, then there could be a `if self.zone.has_zone(z)` which does
```python
     def has_zone(self, zone_name):
         return self.get_zone(zone_name, required=False) is not None
```

but then we have 4 such basic operations... `check_zone()`, `get_zone()`, `get_zones()`, `has_zone()`.

do you like that better?

Reply:
hm. seems the common naming pattern is `query_*()`.

that seems an unsual term, for a "has()" check.

also, it's not used consistently, e.g. `query_masquerade()` is more like `get_masquerade()`.

I don't mind. I can also add a `query_zone()` for the meaning of `has_zone()`.

but maybe the fact that we have such helper functions with unclear names, indicates we shouldn't add more of them. `if self.zone.get_zone(zone_name, required=False)` can be a fine replacement...

Reply:
There is now a `has_{zone,policy,ipset}()`, which is used instead of `get_{zone,policy,ipset}(name, required=False)`.